File: plugins/capacitor-widget-bridge/android/src/main/java/com/stjarndag/widgetbridge/WidgetBridgePlugin.java
package com.stjarndag.widgetbridge;

import android.content.Intent;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "WidgetBridge")
public class WidgetBridgePlugin extends Plugin {

    @PluginMethod
    public void configureBinding(PluginCall call) {
        String token = call.getString("bindingToken");
        String childId = call.getString("activeChildId");
        if (token == null || token.isEmpty() || childId == null || childId.isEmpty()) {
            call.reject("bindingToken and activeChildId required");
            return;
        }
        try {
            WidgetBridgeStore.saveBinding(
                getContext(),
                token,
                childId,
                call.getString("viewerMode"),
                call.getString("privacyMode"),
                call.getString("installationId")
            );
            notifyWidgetRefresh();
            call.resolve(new JSObject().put("ok", true));
        } catch (Exception e) {
            call.reject("Failed to store binding", e);
        }
    }

    @PluginMethod
    public void refreshAll(PluginCall call) {
        WidgetBridgeStore.touchRefresh(getContext());
        notifyWidgetRefresh();
        call.resolve(new JSObject().put("ok", true));
    }

    @PluginMethod
    public void clearBindings(PluginCall call) {
        WidgetBridgeStore.clearAll(getContext());
        notifyWidgetRefresh();
        call.resolve(new JSObject().put("ok", true));
    }

    @PluginMethod
    public void notifyChildChanged(PluginCall call) {
        String childId = call.getString("activeChildId");
        if (childId != null && !childId.isEmpty()) {
            WidgetBridgeStore.publicPrefs(getContext())
                .edit()
                .putString("active_child_id", childId)
                .apply();
        }
        WidgetBridgeStore.invalidatePendingAction(getContext());
        notifyWidgetRefresh();
        call.resolve(new JSObject().put("ok", true));
    }

    @PluginMethod
    public void getStatus(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("hasBinding", WidgetBridgeStore.hasBinding(getContext()));
        ret.put("platform", "android");
        var pub = WidgetBridgeStore.publicPrefs(getContext());
        ret.put("installationId", pub.getString("installation_id", null));
        ret.put("activeChildId", pub.getString("active_child_id", null));
        ret.put("viewerMode", pub.getString("viewer_mode", null));
        ret.put("privacyMode", pub.getString("privacy_mode", null));
        ret.put("lastRefreshAt", pub.getString("last_refresh_at", null));
        call.resolve(ret);
    }

    private void notifyWidgetRefresh() {
        Intent intent = new Intent("com.stjarndag.widget.REFRESH");
        intent.setPackage(getContext().getPackageName());
        getContext().sendBroadcast(intent);
    }
}

File: plugins/capacitor-widget-bridge/ios/Plugin/WidgetBridgePlugin.swift
import Foundation
import Capacitor
import WidgetKit

@objc(WidgetBridgePlugin)
public class WidgetBridgePlugin: CAPPlugin {

    @objc func configureBinding(_ call: CAPPluginCall) {
        guard let token = call.getString("bindingToken"), !token.isEmpty else {
            call.reject("bindingToken required")
            return
        }
        guard let childId = call.getString("activeChildId"), !childId.isEmpty else {
            call.reject("activeChildId required")
            return
        }
        do {
            try WidgetBridgeStore.saveBinding(
                token: token,
                activeChildId: childId,
                viewerMode: call.getString("viewerMode"),
                privacyMode: call.getString("privacyMode"),
                installationId: call.getString("installationId")
            )
            if #available(iOS 14.0, *) {
                WidgetCenter.shared.reloadAllTimelines()
            }
            call.resolve(["ok": true])
        } catch {
            call.reject("Failed to store binding", nil, error)
        }
    }

    @objc func refreshAll(_ call: CAPPluginCall) {
        WidgetBridgeStore.defaultsRefreshTimestamp()
        if #available(iOS 14.0, *) {
            WidgetCenter.shared.reloadAllTimelines()
        }
        call.resolve(["ok": true])
    }

    @objc func clearBindings(_ call: CAPPluginCall) {
        WidgetBridgeStore.clearAll()
        if #available(iOS 14.0, *) {
            WidgetCenter.shared.reloadAllTimelines()
        }
        call.resolve(["ok": true])
    }

    @objc func notifyChildChanged(_ call: CAPPluginCall) {
        if let childId = call.getString("activeChildId"), !childId.isEmpty {
            UserDefaults(suiteName: WidgetBridgeStore.appGroupId)?.set(childId, forKey: "active_child_id")
        }
        WidgetBridgeStore.invalidatePendingAction()
        if #available(iOS 14.0, *) {
            WidgetCenter.shared.reloadAllTimelines()
        }
        call.resolve(["ok": true])
    }

    @objc func getStatus(_ call: CAPPluginCall) {
        call.resolve(WidgetBridgeStore.status())
    }
}

extension WidgetBridgeStore {
    static func defaultsRefreshTimestamp() {
        defaults?.set(ISO8601DateFormatter().string(from: Date()), forKey: "last_refresh_at")
        defaults?.synchronize()
    }
}
